engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/IntermediateThrowEventProcessor.java

```diff
@@ -40,43 +50,143 @@ public Class<ExecutableIntermediateThrowEvent> getType() {
+          .ifRightOrLeft(
+              completed -> stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed),
+              failure -> incidentBehavior.createIncident(failure, completing));
+    }
+  }
+
+  private class MessageIntermediateThrowEventBehavior implements IntermediateThrowEventBehavior {
+
+    @Override
+    public boolean isSuitableForEvent(final ExecutableIntermediateThrowEvent element) {
+      return element.isMessageThrowEvent();
+    }
+
+    @Override
+    public void onActivate(
+        final ExecutableIntermediateThrowEvent element, final BpmnElementContext activating) {
+      if (element.getJobWorkerProperties() != null) {
+        variableMappingBehavior
+            .applyInputMappings(activating, element)
+            .flatMap(ok -> jobBehavior.createNewJob(activating, element))
+            .ifRightOrLeft(
+                ok -> stateTransitionBehavior.transitionToActivated(activating),
+                failure -> incidentBehavior.createIncident(failure, activating));
+      }
+    }
+
+    @Override
+    public void onComplete(
+        final ExecutableIntermediateThrowEvent element, final BpmnElementContext completing) {
+      variableMappingBehavior
+          .applyOutputMappings(completing, element)
+          .flatMap(ok -> stateTransitionBehavior.transitionToCompleted(element, completing))
+          .ifRightOrLeft(
+              completed -> stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed),
+              failure -> incidentBehavior.createIncident(failure, completing));
+    }
+
+    @Override
+    public void onTerminate(
+        final ExecutableIntermediateThrowEvent element, final BpmnElementContext terminating) {
+      if (element.getJobWorkerProperties() != null) {
+        jobBehavior.cancelJob(terminating);
+      }
+    }
+  }
+
+  private class LinkIntermediateThrowEventBehavior implements IntermediateThrowEventBehavior {
+    @Override
+    public boolean isSuitableForEvent(final ExecutableIntermediateThrowEvent element) {
+      return element.isLinkThrowEvent();
+    }
+
+    @Override
+    public void onActivate(
+        final ExecutableIntermediateThrowEvent element, final BpmnElementContext activating) {
+      final var activated = stateTransitionBehavior.transitionToActivated(activating);
+      stateTransitionBehavior.completeElement(activated);
+    }
+
+    @Override
+    public void onComplete(
+        final ExecutableIntermediateThrowEvent element, final BpmnElementContext completing) {
+      final var link = element.getLink();
+      variableMappingBehavior
+          .applyOutputMappings(completing, element)
+          .flatMap(ok -> stateTransitionBehavior.transitionToCompleted(element, completing))
+          .ifRightOrLeft(
+              completed ->
+                  elementActivationBehavior.activateElement(
+                      completing.getRecordValue(), link.getCatchEventElement()),
```

Comment:
:x: We should use `BpmnStateTransitionBehavior#activateElementInstanceInFlowScope` instead.

The `ElementActivationBehavior` should be used only for process instance modification or to start a process instance anywhere.

Unfortunately, this is not very explicit in the code. :sweat_smile: